perf/src/bin/perf_client.rs

```diff
@@ -261,41 +317,59 @@ async fn request_uni(
     acceptor: UniAcceptor,
     upload: u64,
     download: u64,
-    stats: &mut RequestStats,
+    stats: Arc<Mutex<Stats>>,
+    request_stats: &mut RequestStats,
 ) -> Result<()> {
-    request(send, upload, download, stats).await?;
+    request(send, upload, download, stats.clone(), request_stats).await?;
     let recv = {
         let mut guard = acceptor.0.lock().await;
         guard
             .next()
             .await
             .ok_or_else(|| anyhow::anyhow!("End of stream"))
     }??;
-    drain_stream(recv, stats).await?;
+    drain_stream(recv, stats, request_stats).await?;
     Ok(())
 }
 
 async fn request(
     mut send: quinn::SendStream,
     mut upload: u64,
     download: u64,
-    stats: &mut RequestStats,
+    stats: Arc<Mutex<Stats>>,
+    request_stats: &mut RequestStats,
 ) -> Result<()> {
-    stats.upload_start = Some(Instant::now());
+    request_stats.upload_start = Some(Instant::now());
     send.write_all(&download.to_be_bytes()).await?;
 
+    let send_stream_stats = Arc::new(StreamStats::new(send.id(), true));
+
+    if upload > 0 {
+        stats
+            .lock()
+            .unwrap()
+            .stream_stats
+            .get_mut()
+            .unwrap()
+            .push(send_stream_stats.clone());
+    }
+
```

Comment:
seems a bit cleaner to just bail?

```suggestion
    if upload == 0 {
        send.finish().await?;
        return Ok(());
    }
    let send_stream_stats = Arc::new(StreamStats::new(send.id(), true));

    stats
        .lock()
        .unwrap()
        .stream_stats
        .get_mut()
        .unwrap()
         .push(send_stream_stats.clone());
```